### PR TITLE
Return IP addresses as endpoints instead of hostnames

### PR DESCRIPTION
**Pull request by shayancanonical**
# Issue
[DPE-1174](https://warthogs.atlassian.net/browse/DPE-1174)

The `endpoint` field of the `mysql_client` relation is expecting IP addresses (so that external consumers can connect to the database). However, MySQL Group Replication is storing hostnames as member addresses.

# Solution
Convert the hostnames to IP addresses for the `endpoint` field.

```
shayan@wellington:~ (07:42:47)  $ juju run-action data-integrator/0 get-credentials --wait --format=y
aml
unit-data-integrator-0:
  UnitId: data-integrator/0
  id: "6"
  results:
    mysql:
      endpoints: 10.201.101.31:3306
      password: uCWa9MCBrMUmZRM8Ua7j1T63
      read-only-endpoints: 10.201.101.7:3306
      username: relation-9
      version: 8.0.32-0ubuntu0.22.04.2
    ok: "True"
  status: completed
  timing:
    completed: 2023-02-04 12:42:58 +0000 UTC
    enqueued: 2023-02-04 12:42:53 +0000 UTC
    started: 2023-02-04 12:42:58 +0000 UTC
```

# Context
We are using the `socket` python lib to convert hostnames to IP addresses. `socket.gethostbyname` handles IP addresses correctly. Since the member addresses are used by Group Replication to communicate with one another, the DNS resolution of the member address (if it is a hostname) is going to work.

```
>>> import socket                                                                                                                                                                                          
>>> socket.gethostbyname('juju-1aa21d-0')                                                                                                                                                                  
'10.201.101.135'                                                                                                                                                                                           
>>> socket.gethostbyname('10.201.101.135')                                                                                                                                                                 
'10.201.101.135'                                                                                                                                                                                           
>>> socket.gethostbyname('blah')                                                                                                                                                                           
Traceback (most recent call last):                                                                                                                                                                         
  File "<stdin>", line 1, in <module>                                                                                                                                                                      
socket.gaierror: [Errno -3] Temporary failure in name resolution                                                                                                                                           
>>> socket.gethostbyname('juju-1aa21d-0:3306')                                                                                                                                                             
Traceback (most recent call last):                                                                                                                                                                         
  File "<stdin>", line 1, in <module>                                                                                                                                                                      
socket.gaierror: [Errno -2] Name or service not known 
>>>
```

# Release Notes
Return IP addresses as endpoints instead of hostnames

[DPE-1174]: https://warthogs.atlassian.net/browse/DPE-1174?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by paulomach — APPROVED** [review]
Tidy

**Review by taurus-forever — APPROVED** [review]
Please contact Paulo Sottovia to uncomment mysql tests in `data-integrator` after merging. Tnx!

File: lib/charms/mysql/v0/mysql.py [hunk 1]
```diff
@@ -743,8 +744,22 @@ def get_cluster_endpoints(self) -> Tuple[str, str]:
 
         topology = status["defaultreplicaset"]["topology"]
 
-        ro_endpoints = {v["address"] for v in topology.values() if v["mode"] == "r/o"}
-        rw_endpoints = {v["address"] for v in topology.values() if v["mode"] == "r/w"}
+        def _get_host_ip(host: str) -> str:
+            try:
+                if ":" in host:
+                    [host, port] = host.split(":")
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply lib/charms/mysql/v0/mysql.py]: addressed in 3d28255b4c6fae84789b4cb4873a4e917bd6d2f7

File: lib/charms/mysql/v0/mysql.py [hunk 2]
```diff
@@ -743,8 +744,22 @@ def get_cluster_endpoints(self) -> Tuple[str, str]:
...
+
+                host_ip = socket.gethostbyname(host)
+                return host_ip if not port else f"{host_ip}:{port}"
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply lib/charms/mysql/v0/mysql.py]: addressed in 3d28255b4c6fae84789b4cb4873a4e917bd6d2f7

**Pull request merged**